samcli/local/lambda_service/local_lambda_invoke_service.py

```diff
@@ -115,47 +118,113 @@ def _construct_error_handling(self):
         self._app.register_error_handler(404, LambdaErrorResponses.generic_path_not_found)
         self._app.register_error_handler(405, LambdaErrorResponses.generic_method_not_allowed)
 
-    def _invoke_request_handler(self, function_name):
+    def _invoke_async_request(self, function_name, request_data, stdout, stderr):
         """
-        Request Handler for the Local Lambda Invoke path. This method is responsible for understanding the incoming
-        request and invoking the Local Lambda Function
+        Asynchronous Request Handler for the Local Lambda Invoke path. Flask response is returned immediately while the
+        lambda function is run in a separate thread.
 
         Parameters
         ----------
         function_name str
-            Name of the function to invoke
+            Name of function to invoke
+
+        request_data str
+            Parameters to pass to lambda function
+
+        stdout io.BaseIO
+            Output stream that stdout should be written to
+
+        stderr io.BaseIO
+            Output stream that stderr should be written to
 
         Returns
         -------
-        A Flask Response response object as if it was returned from Lambda
+        A flask Response response object as if it was returned from Lambda
         """
-        flask_request = request
+        request_id = uuid4()
 
-        request_data = flask_request.get_data()
+        thread = threading.Thread(target=self._invoke_sync_request,
+                                  args=(function_name, request_data),
+                                  kwargs={'stdout': stdout, 'stderr': stderr})
+        thread.daemon = True
+        thread.start()
 
-        if not request_data:
-            request_data = b'{}'
+        LOG.debug('Async invocation: %s, requestId=%s', function_name, request_id)
```

Comment:
So I just checked `awslambda.js` in the docker container and it looks like `awsRequestId` is set from `invokeId`. `invokeId` is set to `uuid()`. If we could just allow an environment variable to overwrite this value then we could pass it from SAM CLI. I don't know where the code for `awslambda.js` is coming from otherwise I'd make this change myself:

```
var fs = require('fs')
var crypto = require('crypto')

var HANDLER = process.argv[2] || process.env.AWS_LAMBDA_FUNCTION_HANDLER || process.env._HANDLER || 'index.handler'
var EVENT_BODY = process.argv[3] || process.env.AWS_LAMBDA_EVENT_BODY ||
  (process.env.DOCKER_LAMBDA_USE_STDIN && fs.readFileSync('/dev/stdin', 'utf8')) || '{}'

var FN_NAME = process.env.AWS_LAMBDA_FUNCTION_NAME || 'test'
var VERSION = process.env.AWS_LAMBDA_FUNCTION_VERSION || '$LATEST'
var MEM_SIZE = process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE || '1536'
var TIMEOUT = process.env.AWS_LAMBDA_FUNCTION_TIMEOUT || '300'
var REGION = process.env.AWS_REGION || process.env.AWS_DEFAULT_REGION || 'us-east-1'
var ACCOUNT_ID = process.env.AWS_ACCOUNT_ID || randomAccountId()
var ACCESS_KEY_ID = process.env.AWS_ACCESS_KEY_ID || 'SOME_ACCESS_KEY_ID'
var SECRET_ACCESS_KEY = process.env.AWS_SECRET_ACCESS_KEY || 'SOME_SECRET_ACCESS_KEY'
var SESSION_TOKEN = process.env.AWS_SESSION_TOKEN
var INVOKED_ARN = process.env.AWS_LAMBDA_FUNCTION_INVOKED_ARN || arn(REGION, ACCOUNT_ID, FN_NAME)
var INVOKE_ID = process.env.AWS_LAMBDA_INVOKE_ID || uuid() // TODO: add this line

function consoleLog(str) {
  process.stderr.write(formatConsole(str))
}

function systemLog(str) {
  process.stderr.write(formatSystem(str) + '\n')
}

function systemErr(str) {
  process.stderr.write(formatErr(str) + '\n')
}

function handleResult(resultStr, cb) {
  if (!process.stdout.write('\n' + resultStr + '\n')) {
    process.stdout.once('drain', cb)
  } else {
    process.nextTick(cb)
  }
}

// Don't think this can be done in the Docker image
process.umask(2)

process.env.AWS_LAMBDA_FUNCTION_NAME = FN_NAME
process.env.AWS_LAMBDA_FUNCTION_VERSION = VERSION
process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE = MEM_SIZE
process.env.AWS_LAMBDA_LOG_GROUP_NAME = '/aws/lambda/' + FN_NAME
process.env.AWS_LAMBDA_LOG_STREAM_NAME = new Date().toISOString().slice(0, 10).replace(/-/g, '/') +
  '/[' + VERSION + ']' + crypto.randomBytes(16).toString('hex')
process.env.AWS_REGION = REGION
process.env.AWS_DEFAULT_REGION = REGION
process.env._HANDLER = HANDLER

var OPTIONS = {
  initInvokeId: uuid(),
  invokeId: INVOKE_ID, // TODO: update this
  handler: HANDLER,
  suppressInit: true,
  credentials: {
    key: ACCESS_KEY_ID,
    secret: SECRET_ACCESS_KEY,
    session: SESSION_TOKEN,
  },
  eventBody: EVENT_BODY,
  contextObjects: {
    // clientContext: '{}',
    // cognitoIdentityId: undefined,
    // cognitoPoolId: undefined,
  },
  invokedFunctionArn: INVOKED_ARN,
}
```